3.13/specializer.md

```diff
@@ -0,0 +1,289 @@
+ # Minimal bookkeeping, no frame creation
+_PRE_INLINE(7, 0, 39)
+_SET_FRAME_NAMES(0, 0, 93831592813424)
+...
+_POST_INLINE(4, 0, 18446744073709551615)
+_SET_FRAME_NAMES(0, 0, 93831592813424)
+```
+
+
+#### Locals interleaving
+
+By interleaving the locals of the new call with the stack of the old
+function frame, we can achieve zero argument copying.
+
+Before a function call, the stack looks like this:
+
+| callable | self_or_null | arg1 | arg2              |
+|----------|--------------|------|-------------------|
+|          |              |      | ^ (stack pointer) |
+
+Notice that the arguments are already layed out in how the locals
+of the new "frame" will be. Thus, we just need to expand the stack
+to accomodate for more locals. Finally, since we are now pointing
+into the current stack as our locals, we offset all `LOAD_FAST`
+and `STORE_FAST` into the current stack.
+
+| callable | self_or_null                              | arg1 | arg2 |                   |
+|----------|-------------------------------------------|------|------|-------------------|
+|          | ^ `LOAD_FAST 0 + offset` of inlined frame |      |      | ^ (stack pointer) |
+
+Finally, we must update the attribute names' tuple that the current 
+frame points to, so that the inlined call's attribute loads load
+the correct attribute names. This is just a simple assignment.
+
+#### Compatiblity
+We aim for 100% compatibility with all existing CPython code, and
+all uses of the C API. We do not claim compatiblity with native
+stack/frame profilers that do not go through the C API.
+
+The core idea is that in the happy case we have no frames, and
+enjoy inlined functions. In the not-so-happy case, such as tracebacks,
+deoptimizations, ``sys._getframe``, or ``locals()``, we
+can reconstruct all information we need on-demand. We do this by storing
+reconstruction information at the end of a trace, and keep a pointer
+to it in the current frame. We update the pointer in the `_PRE_INLINE`
+and `_POST_INLINE` instructions. This makes the happy case much faster.
+
+As a proof of concept, the author has already implemented full function
+inlining with reconstruction in their branch of the tracer. It has
+full compatibility with tracebacks and deoptimizations. Thus
+proving that frame reconstruction is indeed feasible.
+
+### Object creation inlining
+
+Python object creation currently
+requires the pushing and popping of at least two "fake frames". One for
+a wrapper around `__init__`, and one for `__init__` itself.
+
+By applying the same ideas in true function inlining, we can eliminate
+all function frame require for object creation.
+
+This is not to be confused with scalar replacement.
+
+### Value numbering
+
+To remove redundant operations on things that are functionally equivalent,
+we should implement symbolic terms via value numbering. For example, in
+the following code:
+```
+x = y
+# This has _GUARD_BOTH_INT
+y + y
+# should no longer have _GUARD_BOTH_INT
+x + x
+```
```

Comment:
It is worth noting that value numbering allows us to perform abstract interpretation on the stack based micro-ops and still get this effect. Without value numbering, we would be able to infer almost nothing, as values on the stack are so empheral.

Reply:
Yes. That's why the IR symbolic terms can be seen as an extension of value numbering.